sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -332,6 +333,62 @@ private SpecializedBlobClientBuilder prepareBuilder() {
         }
     }
 
+
+    /**
+     * Creates a new block blob. By default this method will not overwrite an existing blob.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.BlobAsyncClient.upload#BinaryData}
+     *
+     * @param data The data to write to the blob.
+     * @return A reactive response containing the information of the uploaded block blob.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<BlockBlobItem> upload(BinaryData data) {
+        try {
+            return upload(data, false);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Creates a new block blob, or updates the content of an existing block blob.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * {@codesnippet com.azure.storage.blob.BlobAsyncClient.upload#BinaryData-boolean}
+     *
+     * @param data The data to write to the blob.
+     * @param overwrite Whether or not to overwrite, should the blob already exist.
+     * @return A reactive response containing the information of the uploaded block blob.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<BlockBlobItem> upload(BinaryData data,
+                                      boolean overwrite) {
+        try {
+            Mono<Void> overwriteCheck;
+            BlobRequestConditions requestConditions;
+
+            if (overwrite) {
+                overwriteCheck = Mono.empty();
+                requestConditions = null;
+            } else {
+                overwriteCheck = exists().flatMap(exists -> exists
+                    ? monoError(logger, new IllegalArgumentException(Constants.BLOB_ALREADY_EXISTS))
+                    : Mono.empty());
+                requestConditions = new BlobRequestConditions().setIfNoneMatch(Constants.HeaderConstants.ETAG_WILDCARD);
+            }
+
+            return overwriteCheck
+                .then(uploadWithResponse(Flux.just(ByteBuffer.wrap(data.toBytes())), null, null, null, null,
```

Comment:
`data.toBytes()` will perform an array copy, which may be fine for now, but do we want to look into another API such as `toReadOnlyByteBuffer()`? @srnagar

Reply:
Yeah. I was disappointed to not find a way to avoid this copy. It would be great improvement.

Reply:
@alzimmermsft We should look into adding `data.toFlux()` method in BinaryData as we discussed offline to prevent array copy.